### PR TITLE
fix: modal widget issues

### LINKED ISSUE #7628: [Bug] After disabling quick dismiss “on close “ property do not work
### Steps to reproduce the behaviour:

1. Drag and drop a Modal widget
2. Open property pane 
3. Disable Quick dismiss
4. Now add any action on "On Close" property
5. Observe that clicking on close icon do not perform on close action

[LOOM DEMO](https://www.loom.com/share/e1a393d0e0124d3285917dd5301f3ee4)

### Important Details

- Version: [Cloud / Self-Hosted vx.x]
- OS: [e.g.MacOSX]
- Environment [production, release, deploy preview]

### LINKED ISSUE #7686: [Bug] While widget grouping on modal, It selects the widgets in the background instead
## Description

While widget grouping on modal, It selects the widgets in the background instead

### Steps to reproduce the behaviour:

1. Go to Editor page
2. drag and drop some widgets on the canvas near the center
3. open a form modal
4. Try cmd drag and grouping the widgets on the modal
5. Observe that the widgets in the background get selected

[LOOM DEMO](https://www.loom.com/share/48a21d9c88144239914979fdc1957228)

### Important Details

- Version: [Cloud / Self-Hosted vx.x]
- OS: [e.g.MacOSX]
- Environment [production, release, deploy preview]

### LINKED ISSUE #7726: [Bug] App header flickers in View mode when modal is open
## Description

App header flickers in View mode when modal is open

### Steps to reproduce the behaviour:

Add steps to reproduce this behaviour, include console / network logs & screenshots

1. Go to an application in View mode
2. Open a modal built from modal widget
3. hover on app header
4. Observe it flickering

### Important Details

- Version: [Cloud / Self-Hosted vx.x]
- OS: [e.g.MacOSX]
- Environment [production, release, deploy preview]

### LINKED ISSUE #7797: [Bug]Select widgets：the drop-down content shows that the text flashes
## Description

[What happened]
when Generate from Data，get a page。
1.Add a drop-down box to select the input content, the content of the down content comes from mysql；The drop-down box appears in two places：Update Form and Insert Form；
2. The content of the drop-down box in Update Form is displayed normally，but the content of the drop-down box in Insert Form is abnormal。The content in the red box is flashing, and the content of the drop-down box cannot be selected。

y
### Steps to reproduce the behaviour:

Add steps to reproduce this behaviour, include console / network logs & screenshots

### Important Details

- Version: [Cloud / Self-Hosted vx.x]
- OS: [e.g.MacOSX]
- Environment [production, release, deploy preview]

### PR DESCRIPTION
**Pull request by rahulramesha**
## Description
Modal widget fixes for,
flickering issues with app header and dropdown in view mode,
widget grouping inside modal,
onClose to be trigerred on close buttton,

Fixes #7686 
Fixes #7726 
Fixes #7628
Fixes #7797 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Manual UI verification

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: modalwidgetgrouping 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.89 **(-0.02)** | 37.02 **(0)** | 33.88 **(-0.02)** | 55.52 **(-0.01)**
 :red_circle: | app/client/src/widgets/ModalWidget/component/index.tsx | 20.83 **(-5.03)** | 13.33 **(3.57)** | 0 **(0)** | 22.39 **(-4.4)**
 :green_circle: | app/client/src/widgets/ModalWidget/widget/index.tsx | 38.37 **(0.44)** | 0 **(0)** | 17.65 **(0)** | 37.8 **(0.45)**</details>

### CODE REVIEW

File: app/client/src/widgets/ModalWidget/component/index.tsx [hunk 1]
```diff
@@ -140,6 +138,15 @@ export default function ModalComponent(props: ModalComponentProps) {
     return omit(allHandles, get(props, "disabledResizeHandles", []));
   }, [props]);
 
+  useEffect(() => {
+    return () => {
+      // handle modal close events when this component unmounts
+      // will be called in all cases :-
```
Comment (riodeuno) [inline app/client/src/widgets/ModalWidget/component/index.tsx:144]: 💯

File: app/client/src/widgets/ModalWidget/component/index.tsx [hunk 2]
```diff
@@ -191,44 +198,32 @@ export default function ModalComponent(props: ModalComponentProps) {
...
-        top={props.top}
-        width={props.width}
-        zIndex={props.zIndex !== undefined ? props.zIndex : Layers.modalWidget}
+      <Overlay
+        canEscapeKeyClose={props.canEscapeKeyClose}
+        canOutsideClickClose={props.canOutsideClickClose}
+        className={props.overlayClassName}
+        enforceFocus={false}
+        hasBackdrop={
+          props.hasBackDrop !== undefined ? !!props.hasBackDrop : true
```
Comment (riodeuno) [inline app/client/src/widgets/ModalWidget/component/index.tsx:219]: Shall we use `useMemo` here? Will it help?

**Review by riodeuno — APPROVED** [review]

**Review by somangshu — APPROVED** [review]
Patched, Fixes the issue on deploy preview. Approving to run tests,

**Review by riodeuno — APPROVED** [review]

**Comment by rahulramesha** [comment]
Merging this to release

**Comment by bohan0** [comment]
I still seem to be getting flickering on dropdowns as of right now (26/09/21 23:37 GMT). I am using the cloud version - how do I check if it on the latest release that is fixed?

**Comment by rahulramesha** [comment]
@bohan0 Apologies for still facing that issue, the fix has not been pushed to production yet. It will be pushed with coming release, by end of day today, 27/09/21.

**Pull request merged**